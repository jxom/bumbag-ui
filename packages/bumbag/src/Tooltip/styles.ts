import { css, cssClass } from '../styled';
import { borderRadius, fontSize, fontWeight, palette, space, theme } from '../utils';
import { getAnimatedAttributes } from '../Modal/styles';

export const TooltipContent = (styleProps) => cssClass`
  background: black;
  border-radius: ${borderRadius('1')(styleProps)};
  color: ${palette('white')(styleProps)};
  opacity: 0.8;
  hyphens: auto;
  font-size: ${fontSize('100')(styleProps)}rem;
  font-weight: ${fontWeight('normal')(styleProps)};
  padding: ${space(1)(styleProps)}rem ${space(2)(styleProps)}rem;

  ${getPlacementAttributes(styleProps)}

  & {
    ${theme(styleProps.themeKey, `styles.base`)(styleProps)};
  }
`;

export const TooltipReference = (styleProps) => cssClass`
  & {
    ${theme(styleProps.themeKey, `styles.base`)(styleProps)};
  }
`;

export const TooltipArrow = (styleProps) => cssClass`
  display: grid;
  overflow: hidden;

  & > svg {
    background-color: transparent;

    & .stroke {
      fill: black;
    }
    & .fill {
      fill: black;
    }
  }

  & {
    ${theme(styleProps.themeKey, `styles.base`)(styleProps)};
  }
`;

export const getPlacementAttributes = (styleProps) => {
  const placementAttributes: { [key: string]: any } = {
    // @ts-ignore
    top: css`
      ${getAnimatedAttributes({
        prevTransformValue: styleProps.unstable_popoverStyles.transform,
        transformX: '0px',
        transformY: '0px',
        defaultSlide: 'bottom',
        defaultExpand: 'bottom',
        slideOffset: `${styleProps.gutter || '10'}px`,
      })(styleProps)};

      & {
        ${theme(styleProps.themeKey, `styles.placements.top`)(styleProps)};
      }
    `,
    // @ts-ignore
    bottom: css`
      ${getAnimatedAttributes({
        prevTransformValue: styleProps.unstable_popoverStyles.transform,
        transformX: '0px',
        transformY: '0px',
        defaultSlide: 'top',
        defaultExpand: 'top',
        slideOffset: `${styleProps.gutter || '10'}px`,
      })(styleProps)};

      & {
        ${theme(styleProps.themeKey, `styles.placements.bottom`)(styleProps)};
      }
    `,
    // @ts-ignore
    left: css`
      ${getAnimatedAttributes({
        prevTransformValue: styleProps.unstable_popoverStyles.transform,
        transformX: '0px',
        transformY: '0px',
        defaultSlide: 'right',
        defaultExpand: 'right',
        slideOffset: `${styleProps.gutter || '10'}px`,
      })(styleProps)};

      & {
        ${theme(styleProps.themeKey, `styles.placements.left`)(styleProps)};
      }
    `,
    // @ts-ignore
    right: css`
      ${getAnimatedAttributes({
        prevTransformValue: styleProps.unstable_popoverStyles.transform,
        transformX: '0px',
        transformY: '0px',
        defaultSlide: 'left',
        defaultExpand: 'left',
        slideOffset: `${styleProps.gutter || '10'}px`,
      })(styleProps)};

      & {
        ${theme(styleProps.themeKey, `styles.placements.right`)(styleProps)};
      }
    `,
    // @ts-ignore
    'top-start': css`
      ${getAnimatedAttributes({
        prevTransformValue: styleProps.unstable_popoverStyles.transform,
        transformX: '0px',
        transformY: '0px',
        defaultSlide: 'bottom',
        defaultExpand: 'bottom',
        slideOffset: `${styleProps.gutter || '10'}px`,
      })(styleProps)};

      & {
        ${theme(styleProps.themeKey, `styles.placements.topStart`)(styleProps)};
      }
    `,
    // @ts-ignore
    'top-end': css`
      ${getAnimatedAttributes({
        prevTransformValue: styleProps.unstable_popoverStyles.transform,
        transformX: '0px',
        transformY: '0px',
        defaultSlide: 'bottom',
        defaultExpand: 'bottom',
        slideOffset: `${styleProps.gutter || '10'}px`,
      })(styleProps)};

      & {
        ${theme(styleProps.themeKey, `styles.placements.topEnd`)(styleProps)};
      }
    `,
    // @ts-ignore
    'bottom-end': css`
      ${getAnimatedAttributes({
        prevTransformValue: styleProps.unstable_popoverStyles.transform,
        transformX: '0px',
        transformY: '0px',
        defaultSlide: 'top',
        defaultExpand: 'top',
        slideOffset: `${styleProps.gutter || '10'}px`,
      })(styleProps)};

      & {
        ${theme(styleProps.themeKey, `styles.placements.bottomEnd`)(styleProps)};
      }
    `,
    // @ts-ignore
    'bottom-start': css`
      ${getAnimatedAttributes({
        prevTransformValue: styleProps.unstable_popoverStyles.transform,
        transformX: '0px',
        transformY: '0px',
        defaultSlide: 'top',
        defaultExpand: 'top',
        slideOffset: `${styleProps.gutter || '10'}px`,
      })(styleProps)};

      & {
        ${theme(styleProps.themeKey, `styles.placements.bottomStart`)(styleProps)};
      }
    `,
  };
  return css`
    ${placementAttributes[styleProps.placement || 'center']};
  `;
};
