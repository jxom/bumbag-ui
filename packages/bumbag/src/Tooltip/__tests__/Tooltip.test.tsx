import * as React from 'react';
import { renderHook } from '@testing-library/react-hooks';
import { Tooltip } from '../Tooltip';
import render from '../../utils/_tests/render';

describe('props', () => {
  it('should render correctly', () => {
    const { container } = render(
      <Tooltip content="This is content" baseId="test">
        Hello world
      </Tooltip>
    );
    expect(container.firstChild).toMatchSnapshot();
  });

  it('should render correctly with CSS props', () => {
    const { container } = render(
      <Tooltip color="primary" content="This is content" baseId="test">
        Hello world
      </Tooltip>
    );
    expect(container.firstChild).toMatchSnapshot();
  });
});

describe('composition', () => {
  describe('as', () => {
    it('should render correctly', () => {
      const { container } = render(
        <Tooltip use="p" content="This is content" baseId="test">
          Hello world
        </Tooltip>
      );
      expect(container.firstChild).toMatchSnapshot();
    });
  });

  describe('hook', () => {
    it('should return with Tooltip props', () => {
      const { result } = renderHook(() => Tooltip.useProps({ content: 'This is content', baseId: 'test' }));
      expect(result.current).toMatchSnapshot();
    });
  });

  describe('render props', () => {
    it('should render correctly', () => {
      const { container } = render(
        <Tooltip content="This is content" baseId="test">
          {(TooltipProps) => <div {...TooltipProps}>Hello world</div>}
        </Tooltip>
      );
      expect(container.firstChild).toMatchSnapshot();
    });
  });
});

describe('overrides', () => {
  it('Tooltip.Content.base should render correctly', () => {
    const { container } = render(
      <Tooltip
        content="This is content"
        overrides={{ Tooltip: { Content: { styles: { base: { backgroundColor: 'red' } } } } }}
        baseId="test"
      >
        hello world
      </Tooltip>
    );
    expect(container.firstChild).toMatchSnapshot();
  });

  it('Tooltip.Reference.base should render correctly', () => {
    const { container } = render(
      <Tooltip
        content="This is content"
        overrides={{ Tooltip: { Reference: { styles: { base: { backgroundColor: 'red' } } } } }}
        baseId="test"
      >
        hello world
      </Tooltip>
    );
    expect(container.firstChild).toMatchSnapshot();
  });
});

describe('theming', () => {
  it('Tooltip.Content.base should render correctly', () => {
    const { container } = render(
      <Tooltip content="This is content" baseId="test">
        hello world
      </Tooltip>,
      {
        // @ts-ignore
        theme: { Tooltip: { Content: { styles: { base: { backgroundColor: 'red' } } } } },
      }
    );
    expect(container.firstChild).toMatchSnapshot();
  });

  it('Tooltip.Reference.base should render correctly', () => {
    const { container } = render(
      <Tooltip content="This is content" baseId="test">
        hello world
      </Tooltip>,
      {
        // @ts-ignore
        theme: { Tooltip: { Reference: { styles: { base: { backgroundColor: 'red' } } } } },
      }
    );
    expect(container.firstChild).toMatchSnapshot();
  });
});

describe('defaultProps', () => {
  it('should render correctly for className', () => {
    const { container } = render(
      <Tooltip content="This is content" baseId="test">
        hello world
      </Tooltip>,
      {
        // @ts-ignore
        theme: { Tooltip: { defaultProps: { className: 'test', color: 'primary' } } },
      }
    );
    expect(container.firstChild).toMatchSnapshot();
  });
});
