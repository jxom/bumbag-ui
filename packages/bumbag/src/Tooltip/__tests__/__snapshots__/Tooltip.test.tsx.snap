// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`composition as should render correctly 1`] = `
.emotion-1 {
  margin: unset;
  padding: unset;
  border: unset;
  background: unset;
  font: unset;
  font-family: inherit;
  font-size: 100%;
  box-sizing: border-box;
}

.emotion-1:focus:not(:focus-visible) {
  outline: none;
}

<p>
  <div
    aria-describedby="test"
    class="emotion-0 emotion-1 emotion-2 fp-Box fp-TooltipReference"
    tabindex="0"
  >
    Hello world
  </div>
</p>
`;

exports[`composition hook should return with Tooltip props 1`] = `
Object {
  "children": <React.Fragment>
    <Tooltip.Reference
      animated={false}
      animating={false}
      baseId="test"
      className="css-1pwrqjj css-1yuzjtp fp-Box"
      hide={[Function]}
      modal={false}
      place={[Function]}
      placement="top"
      setAnimated={[Function]}
      setModal={[Function]}
      setVisible={[Function]}
      show={[Function]}
      stopAnimation={[Function]}
      toggle={[Function]}
      unstable_arrowRef={
        Object {
          "current": null,
        }
      }
      unstable_arrowStyles={Object {}}
      unstable_disclosureRef={
        Object {
          "current": null,
        }
      }
      unstable_idCountRef={
        Object {
          "current": 0,
        }
      }
      unstable_originalPlacement="top"
      unstable_popoverRef={
        Object {
          "current": null,
        }
      }
      unstable_popoverStyles={Object {}}
      unstable_referenceRef={
        Object {
          "current": null,
        }
      }
      unstable_setBaseId={[Function]}
      unstable_setTimeout={[Function]}
      unstable_timeout={0}
      unstable_update={[Function]}
      visible={false}
    />
    <Tooltip.Content
      animated={false}
      animating={false}
      baseId="test"
      hide={[Function]}
      modal={false}
      place={[Function]}
      placement="top"
      setAnimated={[Function]}
      setModal={[Function]}
      setVisible={[Function]}
      show={[Function]}
      stopAnimation={[Function]}
      toggle={[Function]}
      unstable_arrowRef={
        Object {
          "current": null,
        }
      }
      unstable_arrowStyles={Object {}}
      unstable_disclosureRef={
        Object {
          "current": null,
        }
      }
      unstable_idCountRef={
        Object {
          "current": 0,
        }
      }
      unstable_originalPlacement="top"
      unstable_popoverRef={
        Object {
          "current": null,
        }
      }
      unstable_popoverStyles={Object {}}
      unstable_referenceRef={
        Object {
          "current": null,
        }
      }
      unstable_setBaseId={[Function]}
      unstable_setTimeout={[Function]}
      unstable_timeout={0}
      unstable_update={[Function]}
      visible={false}
    >
      This is content
    </Tooltip.Content>
  </React.Fragment>,
}
`;

exports[`composition render props should render correctly 1`] = `
<div>
  Hello world
</div>
`;

exports[`defaultProps should render correctly for className 1`] = `
.emotion-1 {
  margin: unset;
  padding: unset;
  border: unset;
  background: unset;
  font: unset;
  font-family: inherit;
  font-size: 100%;
  box-sizing: border-box;
}

.emotion-1:focus:not(:focus-visible) {
  outline: none;
}

.emotion-3 {
  color: #574feb !important;
}

<div>
  <div
    aria-describedby="test"
    class="emotion-0 emotion-1 emotion-2 emotion-3 test fp-Box fp-TooltipReference"
    tabindex="0"
  >
    hello world
  </div>
</div>
`;

exports[`overrides Tooltip.Content.root should render correctly 1`] = `
.emotion-1 {
  margin: unset;
  padding: unset;
  border: unset;
  background: unset;
  font: unset;
  font-family: inherit;
  font-size: 100%;
  box-sizing: border-box;
}

.emotion-1:focus:not(:focus-visible) {
  outline: none;
}

<div>
  <div
    aria-describedby="test"
    class="emotion-0 emotion-1 emotion-2 fp-Box fp-TooltipReference"
    tabindex="0"
  >
    hello world
  </div>
</div>
`;

exports[`overrides Tooltip.Reference.root should render correctly 1`] = `
.emotion-1 {
  margin: unset;
  padding: unset;
  border: unset;
  background: unset;
  font: unset;
  font-family: inherit;
  font-size: 100%;
  box-sizing: border-box;
}

.emotion-1:focus:not(:focus-visible) {
  outline: none;
}

.emotion-0 {
  background-color: red;
}

<div>
  <div
    aria-describedby="test"
    class="emotion-0 emotion-1 emotion-2 fp-Box fp-TooltipReference"
    tabindex="0"
  >
    hello world
  </div>
</div>
`;

exports[`props should render correctly 1`] = `
.emotion-1 {
  margin: unset;
  padding: unset;
  border: unset;
  background: unset;
  font: unset;
  font-family: inherit;
  font-size: 100%;
  box-sizing: border-box;
}

.emotion-1:focus:not(:focus-visible) {
  outline: none;
}

<div>
  <div
    aria-describedby="test"
    class="emotion-0 emotion-1 emotion-2 fp-Box fp-TooltipReference"
    tabindex="0"
  >
    Hello world
  </div>
</div>
`;

exports[`props should render correctly with CSS props 1`] = `
.emotion-1 {
  margin: unset;
  padding: unset;
  border: unset;
  background: unset;
  font: unset;
  font-family: inherit;
  font-size: 100%;
  box-sizing: border-box;
}

.emotion-1:focus:not(:focus-visible) {
  outline: none;
}

.emotion-3 {
  color: #574feb !important;
}

<div>
  <div
    aria-describedby="test"
    class="emotion-0 emotion-1 emotion-2 emotion-3 fp-Box fp-TooltipReference"
    tabindex="0"
  >
    Hello world
  </div>
</div>
`;

exports[`theming Tooltip.Content.root should render correctly 1`] = `
.emotion-1 {
  margin: unset;
  padding: unset;
  border: unset;
  background: unset;
  font: unset;
  font-family: inherit;
  font-size: 100%;
  box-sizing: border-box;
}

.emotion-1:focus:not(:focus-visible) {
  outline: none;
}

<div>
  <div
    aria-describedby="test"
    class="emotion-0 emotion-1 emotion-2 fp-Box fp-TooltipReference"
    tabindex="0"
  >
    hello world
  </div>
</div>
`;

exports[`theming Tooltip.Reference.root should render correctly 1`] = `
.emotion-1 {
  margin: unset;
  padding: unset;
  border: unset;
  background: unset;
  font: unset;
  font-family: inherit;
  font-size: 100%;
  box-sizing: border-box;
}

.emotion-1:focus:not(:focus-visible) {
  outline: none;
}

.emotion-0 {
  background-color: red;
}

<div>
  <div
    aria-describedby="test"
    class="emotion-0 emotion-1 emotion-2 fp-Box fp-TooltipReference"
    tabindex="0"
  >
    hello world
  </div>
</div>
`;
